### DISCUSSION TITLE
Where to load the initial state of a store (pinia), which calls an API

### DISCUSSION THREAD

**Discussion by ahoiroman** [nuxt/nuxt#16447] [Questions] [answered]
Hi everybody,

I got a Pinia store that contains a list of user-activities. This list is being loaded using a HTTP-request:

```
export  const useActivityStore = defineStore('activity',  () => {

    const {$auth} = useNuxtApp();

    const activities: Ref<ActivityData[]> = ref([]);

    const getActivities = computed(() => {
        if (!activities.value) return [];

        return activities;
    });

    async function fetchActivities() {
        const {
            data: result,
        } = await useFetch(`/api/profile/${$auth.user.id}/activity`);

        activities.value = result.value?.data || [];
    }

    return {getActivities, fetchActivities, activities}
});
```

Now I can consume this inside a component like this:

```
<template>
    <ul v-if="activityStore.getActivities">
        <li v-for="activity in activityStore.activities">
            {{ activity }}
        </li>
    </ul>
</template>
<script setup lang="ts">
import {useActivityStore} from "~/stores/activityStore";

const activityStore = useActivityStore();
activityStore.fetchActivities();
</script>
```

But this would mean, that the `fetchActivities`-method is running every time the component is mounted, right?

But I want this method to run only if the browser refreshes to fetch initial state. So I thought about putting it into my layout file.

But on the other hand: It's not always ensured a layout is used only with pages that have ensured to have a logged-in user.

I could add a simply condition to make sure the user is loggedIn like this:

*layout*
```
<template>
   ...
</template>

<script setup>

const {$auth} = useNuxtApp();
//...

if($auth.isLoggedIn){
    activityStore.fetchActivities();
}
</script>

```

What do you think? What would be the best approach to do that?

**Comment by Diizzayy** [comment, marked as answer]
@ahoiroman I'd suggest placing the logic for initially fetching and populating the state inside a plugin, this can be tweaked to skip the initialization based on your conditions ( auth, etc ) and would ensure that the pertinent data is only ever loaded upon initial load or when manually triggered elsewhere.

```ts
export default defineNuxtPlugin(async () => {
  const activityStore = useActivityStore();

  // conditions here
  await activityStore.fetchActivities();
})
```

Also, I suggest replace the `useFetch` call in the fetchActivities function with `$fetch`, as it's usage is essentially ineffective here whilst possibly leading to unnecessarily duplicated state in pinia and the nuxt payload.
